man/tss2_getrandom.1.md
% tss2_getrandom(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_getrandom**(1) -
# SYNOPSIS

**tss2_getrandom** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_getrandom**(1) - This command uses the TPM to create an array of random bytes.

# OPTIONS

These are the available options:

  * **-n**, **\--numBytes**=_INTEGER_:

    The number of bytes requested by the caller.

  * **-f**, **\--force**:

    Force overwriting the output file.

  * **-o**, **\--data**=_FILENAME_ or _-_ (for stdout):

    The returned random bytes.

  * **\--hex**

    Convert the output data to hex format without a leading "0x".

[common tss2 options](common/tss2-options.md)

# EXAMPLE
```
    tss2_getrandom --numBytes=20 --data=- | hexdump -C
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_getrandom.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "tools/fapi/tss2_template.h"

/* Context struct used to store passed commandline parameters */
static struct cxt {
    size_t  numBytes;
    char   *filename;
    bool    overwrite;
    bool    hex;
} ctx;

/* Parse commandline parameters */
static bool on_option(char key, char *value) {
    switch (key) {
    case 'n': {
        /*N.B. In theory size_t can be unsigned long, which is more than
         * uint32, in practice this will never happen */
        uint32_t i;
        if (!tpm2_util_string_to_uint32 (value, &i) || i == 0) {
            fprintf (stderr, "%s cannot be converted to a positive integer or "\
                "is larger than 2**32 - 1\n", value);
            return false;
        }
        ctx.numBytes = i; /* cast from uint32 to size_t */
        }
        break;
    case 'f':
        ctx.overwrite = true;
        break;
    case 'o':
        ctx.filename = value;
        break;
    case 0:
        ctx.hex = true;
        break;
    }
    return true;
}

/* Define possible commandline parameters */
static bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"numBytes", required_argument, NULL, 'n'},
        {"force"    , no_argument      , NULL, 'f'},
        /* output file */
        {"data"   , required_argument, NULL, 'o'},
        {"hex",          no_argument,       NULL,  0}
    };
    return (*opts = tpm2_options_new ("fn:o:", ARRAY_LEN(topts), topts,
                                      on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
static int tss2_tool_onrun (FAPI_CONTEXT *fctx) {
    /* Check availability of required parameters */
    if (!ctx.filename) {
        fprintf (stderr, "No filename for data was provided, use --data\n");
        return -1;
    }
    if (!ctx.numBytes) {
        fprintf (stderr, "No amount of bytes was provided, use --numBytes\n");
        return -1;
    }

    /* Execute FAPI command with passed arguments */
    uint8_t *data;
    TSS2_RC r = Fapi_GetRandom (fctx, ctx.numBytes, &data);
    if (r != TSS2_RC_SUCCESS) {
        LOG_PERR ("Fapi_GetRandom", r);
        return 1;
    }

    if (ctx.hex) {
        char* str = malloc (ctx.numBytes*2 + 1);
        if (!str) {
            Fapi_Free (data);
            LOG_ERR ("malloc(2) failed: %m\n");
            return 1;
        }
        for (size_t i = 0; i<ctx.numBytes; i++) {
            sprintf(str+i*2,"%02x",data[i]);
        }
        /* Write returned data to file(s) */
        r = open_write_and_close (ctx.filename, ctx.overwrite, str, strlen(str));
        free(str);
    }
    else {
        /* Write returned data to file(s) */
        r = open_write_and_close (ctx.filename, ctx.overwrite, data,
            ctx.numBytes);
    }
    if (r) {
        Fapi_Free (data);
        return 1;
    }

    Fapi_Free (data);
    return 0;
}

TSS2_TOOL_REGISTER("getrandom", tss2_tool_onstart, tss2_tool_onrun, NULL)
